# Lesson 0004 - random forests
In this lesson, we will, again, revisit the iris data set and apply [random forests](https://en.wikipedia.org/wiki/Random_forest) to the task of classification.<br>
As in [lesson 0002](https://github.com/Mathhead/Lessons-in-Machine-Learning/blob/master/lesson_0002_support_vector_machines_classification.ipynb), we assume, that the reader has worked through the previous examples.

In [1]:
import sklearn as sklearn
from sklearn import datasets

data = datasets.load_iris()

print( sklearn.__version__ )

0.19.1


In [2]:
data_x = data.data

data_y = data.target

In [3]:
import numpy as np

print( np.__version__ )

1.14.3


In [4]:
np.random.seed( 1234567890 )

In [5]:
random_integers = np.random.choice( range( 150 ), 100, replace = False )

random_integers = np.sort( random_integers )



train_x = np.zeros( shape = [ 100, 4 ] )

train_y = np.zeros( shape = [ 100 ] )

test_x = np.zeros( shape = [ 50, 4 ] )

test_y = np.zeros( shape = [ 50 ] )



j = 0

k = 0


for i in range( 150 ):
    
    if i == random_integers[ j ]:
        
        train_x[ j, : ] = data_x[ i, : ]
        
        train_y[ j ] = data_y[ i ]
        
        j = j + 1
        
        if j == 100:
            
            j = 0
            
    else:
        
        test_x[ k, : ] = data_x[ i, : ]
        
        test_y[ k ] = data_y[ i ]
        
        k = k + 1
        
        
        
        


mu = np.zeros( shape = [ 4 ] )

sigma = np.zeros( shape = [ 4 ] )



for i in range( 4 ):
    
    mu[ i ] = np.mean( train_x[ :, i ] )
    
    sigma[ i ] = np.std( train_x[ :, i ] )
    
    
for i in range( 4 ):
    
    train_x[ :, i ] = ( train_x[ :, i ] - mu[ i ] ) / sigma[ i ]
    
    test_x[ :, i ] = ( test_x[ :, i ] - mu[ i ] ) / sigma[ i ]

Now we import the [random forest classifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier) from sklearn.

In [6]:
from sklearn.ensemble import RandomForestClassifier as RFC

We will use __RFC__ right out of the box to fit a model __forest__. We will make predictions using __forest__ and store the results in __hit__ which we will use to compute the __accuracy__.

In [7]:
forest = RFC()

forest.fit( train_x, train_y )

hit = ( forest.predict( test_x ) == test_y )

accuracy = 0.0

for i in range( 50 ):
    
    if hit[ i ]:
        
        accuracy = accuracy + 1.0
        
accuracy = accuracy * 2.0

print( 'We reached an accurady of ' + str( accuracy ) + '%' )

We reached an accurady of 96.0%


Since we reached an accuracy of $96\%$, there is hardly any room for fine tuning since we only have $50$ data items, therefore<br>
Class dismissed.